spyder/plugins/ipythonconsole/utils/kernel_handler.py

```diff
@@ -0,0 +1,475 @@
+            if obj is None:
+                continue
+            obj.remove()
+            fn = obj.filename
+            m = re.match(r"(.+_)(\d+)(.[a-z]+)", fn)
+            if m:
+                # Already a replaced file
+                path, n, ext = m.groups()
+                obj.filename = path + str(1 + int(n)) + ext
+                continue
+            m = re.match(r"(.+)(.[a-z]+)", fn)
+            if m:
+                # First replaced file
+                path, ext = m.groups()
+                obj.filename = path + "_1" + ext
+                continue
+            # No extension, should not happen
+            obj.filename += "_1"
+
+    def get_fault_filename(self):
+        """Get faulthandler filename"""
+        if self.fault_obj is None:
+            return
+        return self.fault_obj.filename
+
+    def get_fault_text(self):
+        """Get a fault from a previous session."""
+
+        if self.fault_obj is None:
+            return
+        fault = self.fault_obj.get_contents()
+        if not fault:
+            return
+
+        thread_regex = (
+            r"(Current thread|Thread) "
+            r"(0x[\da-f]+) \(most recent call first\):"
+            r"(?:.|\r\n|\r|\n)+?(?=Current thread|Thread|\Z)")
+        # Keep line for future improvments
+        # files_regex = r"File \"([^\"]+)\", line (\d+) in (\S+)"
+
+        main_re = "Main thread id:(?:\r\n|\r|\n)(0x[0-9a-f]+)"
+        main_id = 0
+        for match in re.finditer(main_re, fault):
+            main_id = int(match.group(1), base=16)
+
+        system_re = ("System threads ids:"
+                     "(?:\r\n|\r|\n)(0x[0-9a-f]+(?: 0x[0-9a-f]+)+)")
+        ignore_ids = []
+        start_idx = 0
+        for match in re.finditer(system_re, fault):
+            ignore_ids = [int(i, base=16) for i in match.group(1).split()]
+            start_idx = match.span()[1]
+        text = ""
+        for idx, match in enumerate(re.finditer(thread_regex, fault)):
+            if idx == 0:
+                text += fault[start_idx:match.span()[0]]
+            thread_id = int(match.group(2), base=16)
+            if thread_id != main_id:
+                if thread_id in ignore_ids:
+                    continue
+                if "wurlitzer.py" in match.group(0):
+                    # Wurlitzer threads are launched later
+                    continue
+                text += "\n" + match.group(0) + "\n"
+            else:
+                try:
+                    pattern = (r".*(?:/IPython/core/interactiveshell\.py|"
+                               r"\\IPython\\core\\interactiveshell\.py).*")
+                    match_internal = next(re.finditer(pattern, match.group(0)))
+                    end_idx = match_internal.span()[0]
+                except StopIteration:
+                    end_idx = None
+                text += "\nMain thread:\n" + match.group(0)[:end_idx] + "\n"
+        return text
+
+
+class CachedKernelMixin:
```

Comment:
Let's move this to `ipythonconsole/widgets/mixins.py` because it's used by `IPythonConsoleWidget`.